File: tourcount/src/main/java/com/wmstein/tourcount/EditIndividualActivity.java
package com.wmstein.tourcount;

import android.Manifest;
import android.annotation.SuppressLint;
import android.content.SharedPreferences;
import android.content.pm.ActivityInfo;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.graphics.drawable.BitmapDrawable;
import android.media.Ringtone;
import android.media.RingtoneManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.StrictMode;
import android.text.Html;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.WindowManager;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.TextView;

import com.google.android.material.snackbar.Snackbar;
import com.wmstein.egm.EarthGravitationalModel;
import com.wmstein.tourcount.database.Count;
import com.wmstein.tourcount.database.CountDataSource;
import com.wmstein.tourcount.database.Individuals;
import com.wmstein.tourcount.database.IndividualsDataSource;
import com.wmstein.tourcount.database.Temp;
import com.wmstein.tourcount.database.TempDataSource;
import com.wmstein.tourcount.widgets.EditIndividualWidget;

import java.io.IOException;
import java.net.URL;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;

/*******************************************************************************************
 * EditIndividualActivity is called from CountingActivity and collects additional info to an 
 * individual's data record
 * Copyright 2016-2018 wmstein, created on 2016-05-15, 
 * last modification an 2021-01-26
 */
public class EditIndividualActivity extends AppCompatActivity implements SharedPreferences.OnSharedPreferenceChangeListener, PermissionsDialogFragment.PermissionsGrantedCallback
{
    private static final String TAG = "TourCountEditIndivAct";
    @SuppressLint("StaticFieldLeak")
    private static TourCountApplication tourCount;
    private SharedPreferences prefs;
    private Individuals individuals;
    private Temp temp;
    private Count counts;
    private LinearLayout individ_area;
    private EditIndividualWidget eiw;

    // The actual data
    private IndividualsDataSource individualsDataSource;
    private TempDataSource tempDataSource;
    private CountDataSource countDataSource;
    private Bitmap bMap;
    private BitmapDrawable bg;
    
    // Preferences
    private boolean buttonSoundPref;
    private String buttonAlertSound;
    private boolean brightPref;    // option for full bright screen
    private boolean screenOrientL; // option for screen orientation
    private boolean metaPref;      // option for reverse geocoding
    private String emailString = ""; // mail address for OSM query

    // Location info handling
    private double latitude, longitude, height, uncertainty;
    LocationService locationService;

    // Permission dispatcher mode modePerm: 
    //  1 = use location service
    //  2 = end location service
    int modePerm;

    private int count_id;
    private int i_id, iAtt;
    private String specName;
    private Boolean sdata; // true: data saved already

    @SuppressLint("SourceLockedOrientationActivity")
    @Override
    protected void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);

        tourCount = (TourCountApplication) getApplication();
        prefs = TourCountApplication.getPrefs();
        prefs.registerOnSharedPreferenceChangeListener(this);
        getPrefs();

        setContentView(R.layout.activity_edit_individual);

        ScrollView individ_screen = findViewById(R.id.editIndividualScreen);

        if (screenOrientL)
        {
            setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE);
        }
        else
        {
            setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
        }

        // Set full brightness of screen
        if (brightPref)
        {
            getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
            WindowManager.LayoutParams params = getWindow().getAttributes();
            params.screenBrightness = 1.0f;
            getWindow().setAttributes(params);
        }

        bMap = tourCount.decodeBitmap(R.drawable.kbackground, tourCount.width, tourCount.height);
        bg = new BitmapDrawable(individ_screen.getResources(), bMap);
        individ_screen.setBackground(bg);

        individ_area = findViewById(R.id.edit_individual);

        // get parameters from CountingActivity
        Bundle extras = getIntent().getExtras();
        if (extras != null)
        {
            count_id = extras.getInt("count_id");
            i_id = extras.getInt("indiv_id");
            specName = extras.getString("SName");
            latitude = extras.getDouble("Latitude");
            longitude = extras.getDouble("Longitude");
            height = extras.getDouble("Height");
            uncertainty = extras.getDouble("Uncert");
            iAtt = extras.getInt("indivAtt");
        }

        sdata = false;
    }

    // Load preferences
    private void getPrefs()
    {
        buttonSoundPref = prefs.getBoolean("pref_button_sound", false);
        buttonAlertSound = prefs.getString("alert_button_sound", null);
        brightPref = prefs.getBoolean("pref_bright", true);
        screenOrientL = prefs.getBoolean("screen_Orientation", false);
        metaPref = prefs.getBoolean("pref_metadata", false);   // use Reverse Geocoding
        emailString = prefs.getString("email_String", "");     // for reliable query of Nominatim service
    }

    @SuppressLint({"LongLogTag", "DefaultLocale"})
    @Override
    protected void onResume()
    {
        super.onResume();

        // Get location with permissions check
        modePerm = 1;
        permissionCaptureFragment();

        // clear any existing views
        individ_area.removeAllViews();

        // setup the data sources
        individualsDataSource = new IndividualsDataSource(this);
        individualsDataSource.open();

        // get last found locality from temp 
        tempDataSource = new TempDataSource(this);
        tempDataSource.open();
        temp = tempDataSource.getTemp();

        String sLocality = temp.temp_loc;

        countDataSource = new CountDataSource(this);
        countDataSource.open();

        // set title
        try
        {
            getSupportActionBar().setTitle(specName);
        } catch (NullPointerException e)
        {
            if (MyDebug.LOG)
                Log.e(TAG, "NullPointerException: No species name!");
        }

        individuals = individualsDataSource.getIndividual(i_id);
        counts = countDataSource.getCountById(count_id);

        // display the editable data
        eiw = new EditIndividualWidget(this, null);
        eiw.setWidgetLocality1(getString(R.string.locality));
        eiw.setWidgetLocality2(sLocality);

        eiw.setWidgetZCoord1(getString(R.string.zcoord));
        eiw.setWidgetZCoord2(String.format("%.1f", height));

        eiw.setWidgetStadium1(getString(R.string.stadium));
        switch (iAtt)
        {
        case 1: // ♂♀
        case 2: // ♂
        case 3: // ♀
            eiw.setWidgetStadium2(getString(R.string.stadium_1));
            break;
        case 4: // Pupa
            eiw.setWidgetStadium2(getString(R.string.stadium_2));
            break;
        case 5: // Larva
            eiw.setWidgetStadium2(getString(R.string.stadium_3));
            break;
        case 6: // Egg
            eiw.setWidgetStadium2(getString(R.string.stadium_4));
            break;
        }

        eiw.setWidgetState1(getString(R.string.state));
        eiw.setWidgetState2(individuals.state_1_6);

        eiw.setWidgetCount1(getString(R.string.count1)); // icount
        eiw.setWidgetCount2(1);

        eiw.setWidgetIndivNote1(getString(R.string.note));
        eiw.setWidgetIndivNote2(individuals.notes);

        eiw.setWidgetXCoord1(getString(R.string.xcoord));
        eiw.setWidgetXCoord2(String.format("%.6f", latitude));

        eiw.setWidgetYCoord1(getString(R.string.ycoord));
        eiw.setWidgetYCoord2(String.format("%.6f", longitude));

        individ_area.addView(eiw);
    }

    @Override
    protected void onPause()
    {
        super.onPause();

        if (!sdata)
        {
            saveData();
        }

        // close the data sources
        individualsDataSource.close();
        tempDataSource.close();
        countDataSource.close();

        // Stop location service with permissions check
        modePerm = 2;
        permissionCaptureFragment();
    }

    private boolean saveData()
    {
        buttonSound();

        // save individual data
        // Locality (from reverse geocoding in CountingActivity or manual input) 
        individuals.locality = eiw.getWidgetLocality2();

        // Uncertainty
        if (latitude != 0)
        {
            individuals.uncert = String.valueOf(uncertainty);
        }
        else
        {
            individuals.uncert = "0";
        }

        temp.temp_loc = eiw.getWidgetLocality2();

        // Stadium
        individuals.stadium = eiw.getWidgetStadium2();

        // State_1-6
        int newstate = eiw.getWidgetState2();
        if (newstate >= 0 && newstate < 7)
        {
            individuals.state_1_6 = newstate;
        }
        else
        {
//            Toast.makeText(this, getString(R.string.valState), Toast.LENGTH_SHORT).show();
            showSnackbarRed(getString(R.string.valState));

            return false;
        }

        // number of individuals
        int newcount = eiw.getWidgetCount2();
        if (newcount > 0) // valid newcount
        {
            switch (iAtt)
            {
            case 1:
                //  counts.count = counts.count + newcount - 1; // -1 when CountingActivity already added 1
                counts.count_f1i = counts.count_f1i + newcount;
                individuals.icount = newcount;
                individuals.sex = "-";
                individuals.icategory = 1;
                countDataSource.saveCountf1i(counts);
                break;

            case 2:
                counts.count_f2i = counts.count_f2i + newcount;
                individuals.icount = newcount;
                individuals.sex = "m";
                individuals.icategory = 2;
                countDataSource.saveCountf2i(counts);
                break;
            case 3:
                counts.count_f3i = counts.count_f3i + newcount;
                individuals.icount = newcount;
                individuals.sex = "f";
                individuals.icategory = 3;
                countDataSource.saveCountf3i(counts);
                break;
            case 4:
                counts.count_pi = counts.count_pi + newcount;
                individuals.icount = newcount;
                individuals.sex = "-";
                individuals.icategory = 4;
                countDataSource.saveCountpi(counts);
                break;
            case 5:
                counts.count_li = counts.count_li + newcount;
                individuals.icount = newcount;
                individuals.sex = "-";
                individuals.icategory = 5;
                countDataSource.saveCountli(counts);
                break;
            case 6:
                counts.count_ei = counts.count_ei + newcount;
                individuals.icount = newcount;
                individuals.sex = "-";
                individuals.icategory = 6;
                countDataSource.saveCountei(counts);
                break;
            }

            // Notes
            String newnotes = eiw.getWidgetIndivNote2();
            if (!newnotes.equals(""))
            {
                individuals.notes = newnotes;
            }

            individualsDataSource.saveIndividual(individuals);
        }
        else // newcount is <= 1
        {
            
//            Toast.makeText(this, getString(R.string.warnCount), Toast.LENGTH_SHORT).show();
            showSnackbarRed(getString(R.string.warnCount));
            return false; // forces input newcount > 0
        }
        
        tempDataSource.saveTempLoc(temp);
        sdata = true;

        return true;
    }

    private void showSnackbarRed(String str)
    {
        View view = findViewById(R.id.editIndividualScreen);
        Snackbar sB = Snackbar.make(view, Html.fromHtml("<font color=\"#ff0000\"><b>" +  str + "</font></b>"), Snackbar.LENGTH_LONG);
        TextView tv = sB.getView().findViewById(R.id.snackbar_text);
        tv.setTextAlignment(View.TEXT_ALIGNMENT_CENTER);
        sB.show();
    }
    private void buttonSound()
    {
        if (buttonSoundPref)
        {
            try
            {
                Uri notification;
                if (isNotBlank(buttonAlertSound) && buttonAlertSound != null)
                {
                    notification = Uri.parse(buttonAlertSound);
                }
                else
                {
                    notification = RingtoneManager.getDefaultUri(RingtoneManager.TYPE_NOTIFICATION);
                }
                Ringtone r = RingtoneManager.getRingtone(getApplicationContext(), notification);
                r.play();
            } catch (Exception e)
            {
                if (MyDebug.LOG)
                    Log.e(TAG, "could not play button sound.", e);
            }
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu)
    {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.edit_individual, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item)
    {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();
        if (id == R.id.menuSaveExit)
        {
            if (saveData())
            {
                super.finish();
                // close the data sources
                individualsDataSource.close();
                tempDataSource.close();
                countDataSource.close();
            }
        }
        return super.onOptionsItemSelected(item);
    }

    @SuppressLint("SourceLockedOrientationActivity")
    public void onSharedPreferenceChanged(SharedPreferences prefs, String key)
    {
        ScrollView individ_screen = findViewById(R.id.editIndividualScreen);
        individ_screen.setBackground(null);
        getPrefs();
        if (screenOrientL)
        {
            setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE);
        }
        else
        {
            setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
        }

        bMap = tourCount.decodeBitmap(R.drawable.kbackground, tourCount.width, tourCount.height);
        bg = new BitmapDrawable(individ_screen.getResources(), bMap);
        individ_screen.setBackground(bg);
    }

    @Override
    public void permissionCaptureFragment()
    {
        {
            if (isPermissionGranted())
            {
                switch (modePerm)
                {
                case 1: // get location
                    getLoc();
                    break;
                case 2: // stop location service
                    locationService.stopListener();
                    break;
                }
            }
            else
            {
                if (modePerm == 1)
                    PermissionsDialogFragment.newInstance().show(getSupportFragmentManager(), PermissionsDialogFragment.class.getName());
            }
        }
    }

    // if API level > 23 test for permissions granted
    private boolean isPermissionGranted()
    {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M)
        {
            return ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED
                && ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED;
        }
        else
        {
            // handle permissions for Build.VERSION_CODES < M here
            return true;
        }
    }

    // get the location data
    public void getLoc()
    {
        locationService = new LocationService(this);

        if (locationService.canGetLocation())
        {
            longitude = locationService.getLongitude();
            latitude = locationService.getLatitude();
            height = locationService.getAltitude();
            if (height != 0)
                height = correctHeight(latitude, longitude, height);
            uncertainty = locationService.getAccuracy();
        }

        // get reverse geocoding
        if (locationService.canGetLocation() && metaPref && (latitude != 0 || longitude != 0))
        {
            StrictMode.ThreadPolicy policy = new StrictMode.ThreadPolicy.Builder().permitAll().build();
            StrictMode.setThreadPolicy(policy);

            runOnUiThread(() -> {
                URL url;
                String urlString = "https://nominatim.openstreetmap.org/reverse?email=" + emailString + "&format=xml&lat="
                    + latitude + "&lon=" + longitude + "&zoom=18&addressdetails=1";
                try
                {
                    url = new URL(urlString);
                    RetrieveAddr getXML = new RetrieveAddr(getApplicationContext());
                    getXML.execute(url);
                } catch (IOException e)
                {
                    // do nothing
                }
            });
        }
    }

    // Correct height with geoid offset from EarthGravitationalModel
    private double correctHeight(double latitude, double longitude, double gpsHeight)
    {
        double corrHeight;
        double nnHeight;

        EarthGravitationalModel gh = new EarthGravitationalModel();
        try
        {
            gh.load(this); // load the WGS84 correction coefficient table egm180.txt
        } catch (IOException e)
        {
            return 0;
        }

        // Calculate the offset between the ellipsoid and geoid
        try
        {
            corrHeight = gh.heightOffset(latitude, longitude, gpsHeight);
        } catch (Exception e)
        {
            return 0;
        }

        nnHeight = gpsHeight + corrHeight;
        return nnHeight;
    }
    
    /**
     * Checks if a CharSequence is whitespace, empty ("") or null
     * <p>
     * isBlank(null)      = true
     * isBlank("")        = true
     * isBlank(" ")       = true
     * isBlank("bob")     = false
     * isBlank("  bob  ") = false
     *
     * @param cs the CharSequence to check, may be null
     * @return {@code true} if the CharSequence is null, empty or whitespace
     */
    private static boolean isBlank(final CharSequence cs)
    {
        int strLen;
        if (cs == null || (strLen = cs.length()) == 0)
        {
            return true;
        }
        for (int i = 0; i < strLen; i++)
        {
            if (!Character.isWhitespace(cs.charAt(i)))
            {
                return false;
            }
        }
        return true;
    }

    /**
     * Checks if a CharSequence is not empty (""), not null and not whitespace only.
     * <p>
     * isNotBlank(null)      = false
     * isNotBlank("")        = false
     * isNotBlank(" ")       = false
     * isNotBlank("bob")     = true
     * isNotBlank("  bob  ") = true
     *
     * @param cs the CharSequence to check, may be null
     * @return {@code true} if the CharSequence is
     * not empty and not null and not whitespace
     */
    private static boolean isNotBlank(final CharSequence cs)
    {
        return !isBlank(cs);
    }

}

File: tourcount/src/main/java/com/wmstein/tourcount/EditIndividualActivity.kt
package com.wmstein.tourcount

import android.annotation.SuppressLint
import android.os.Build
import android.os.Bundle
import android.os.VibrationEffect
import android.os.Vibrator
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.ViewGroup.MarginLayoutParams
import android.view.WindowManager
import android.widget.LinearLayout
import android.widget.Toast
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.updateLayoutParams
import com.wmstein.tourcount.TourCountApplication.Companion.heightNN
import com.wmstein.tourcount.TourCountApplication.Companion.lat
import com.wmstein.tourcount.TourCountApplication.Companion.lon
import com.wmstein.tourcount.TourCountApplication.Companion.sLocality
import com.wmstein.tourcount.TourCountApplication.Companion.uncertainty
import com.wmstein.tourcount.Utils.fromHtml
import com.wmstein.tourcount.database.Count
import com.wmstein.tourcount.database.CountDataSource
import com.wmstein.tourcount.database.Individuals
import com.wmstein.tourcount.database.IndividualsDataSource
import com.wmstein.tourcount.widgets.EditIndividualWidget

/*******************************************************************************************
 * EditIndividualActivity is called from CountingActivity and collects additional info to an
 * individual's data record
 *
 * Created by wmstein on 2016-05-15,
 * last modification in Java an 2023-07-09,
 * converted to Kotlin on 2023-07-11,
 * last edited on 2026-05-05
 */
class EditIndividualActivity : AppCompatActivity() {
    private var individuals: Individuals? = null
    private var counts: Count? = null
    private var indivArea: LinearLayout? = null
    private var eiw: EditIndividualWidget? = null

    // The actual data
    private var individualsDataSource: IndividualsDataSource? = null
    private var countDataSource: CountDataSource? = null

    // Preferences
    private val prefs = TourCountApplication.getPrefs()
    private var awakePref = false
    private var brightPref = false // option for full bright screen
    private var buttonSoundPref = false
    private var buttonVibPref = false
    private var metaPref: Boolean = false // use Reverse Geocoding

    // Prepare sound service
    private var soundService: SoundService? = null

    // Prepare vibrator service
    private var vibrator: Vibrator? = null

    // Location info handling
    private var uncert: String? = ""
    private var countId = 0
    private var indivId = 0
    private var indivAttr = 0 // 1 = ♂|♀, 2 = ♂, 3 = ♀, 4 = caterpillar, 5 = pupa, 6 = egg
    private var specName: String? = ""

    // phase123 is true for butterfly (♂|♀, ♂ or ♀), false for egg, caterpillar or pupa
    private var phase123: Boolean? = null
    private var dateStamp: String? = ""
    private var timeStamp: String? = ""
    private var code: String? = ""
    private var mesg: String? = ""

    @SuppressLint("SourceLockedOrientationActivity", "DefaultLocale")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        if (IsRunningOnEmulator.DLOG || BuildConfig.DEBUG)
            Log.i(TAG, "84, onCreate")

        awakePref = prefs.getBoolean("pref_awake", true) // keep screen on
        brightPref = prefs.getBoolean("pref_bright", true)
        buttonSoundPref = prefs.getBoolean("pref_button_sound", false)
        buttonVibPref = prefs.getBoolean("pref_button_vib", false)
        metaPref = prefs.getBoolean("pref_metadata", false) // use Reverse Geocoding

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.VANILLA_ICE_CREAM) // SDK 35+
        {
            enableEdgeToEdge()
        }
        setContentView(R.layout.activity_edit_individual)

        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.editIndividualScreen))
        { v, windowInsets ->
            val insets = windowInsets.getInsets(WindowInsetsCompat.Type.systemBars())
            // Apply the insets as a margin to the view. This solution sets
            // only the bottom, left, and right dimensions, but you can apply whichever
            // insets are appropriate to your layout. You can also update the view padding
            // if that's more appropriate.
            v.updateLayoutParams<MarginLayoutParams> {
                topMargin = insets.top
                leftMargin = insets.left
                bottomMargin = insets.bottom
                rightMargin = insets.right
            }

            // Return CONSUMED if you don't want the window insets to keep passing
            // down to descendant views.
            WindowInsetsCompat.CONSUMED
        }

        // Set full brightness of screen
        if (brightPref) {
            val params = window.attributes
            params.screenBrightness = 1.0f
            window.attributes = params
        }

        if (awakePref)
            window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)

        indivArea = findViewById(R.id.edit_individual)

        // Make (+)-button sound
        if (buttonSoundPref) {
            soundService = SoundService(applicationContext)
            soundService!!.soundPlusButtonSound() // sound for (+)-button
        }

        // Prepare vibrator service
        if (buttonVibPref) {
            vibrator = applicationContext.getSystemService(Vibrator::class.java)
            buttonVib()
        }

        // Get parameters from CountingActivity
        val extras = intent.extras
        if (extras != null) {
            countId = extras.getInt("count_id")
            if (extras.getString("SName") != "")
                specName = extras.getString("SName")
            code = extras.getString("SCode")
            dateStamp = extras.getString("date")
            timeStamp = extras.getString("time")
            indivAttr = extras.getInt("indivAtt")
        }

        uncert = String.format("%f", uncertainty)
    }
    // End of onCreate()

    override fun onResume() {
        super.onResume()

        if (IsRunningOnEmulator.DLOG || BuildConfig.DEBUG)
            Log.i(TAG, "161, onResume")

        // Clear any existing views
        indivArea!!.removeAllViews()

        // Set up the data sources
        individualsDataSource = IndividualsDataSource(this)
        individualsDataSource!!.open()

        countDataSource = CountDataSource(this)
        countDataSource!!.open()

        // Set title
        supportActionBar!!.title = specName

        counts = countDataSource!!.getCountById(countId)

        displayData()
    }
    // End of onResume()

    // Display the editable data
    @SuppressLint("DefaultLocale")
    fun displayData() {
        if (IsRunningOnEmulator.DLOG || BuildConfig.DEBUG) {
            Log.i(TAG, "186, displayData: $sLocality")
        }

        eiw = EditIndividualWidget(this, null)
        eiw!!.setWidgetZCoord1(getString(R.string.zcoord))

        // Height value from GPS_PROVIDER
        if (heightNN != 0.0) {
            // Set string with value from double for heightNN
            eiw!!.setWidgetZCoord2(
                String.format(
                    "%.0f",
                    heightNN
                )
            )
        } else { // No height value from NETWORK_PROVIDER
            // Set "-" as there is no height value
            eiw!!.setWidgetZCoord2("-")
        }
        eiw!!.setWidgetStadium1(getString(R.string.stadium) + ":")
        when (indivAttr) {
            1, 2, 3 -> {
                eiw!!.widgetStadium2 = getString(R.string.stadium_1)
                phase123 = true
            }

            4 -> { // Pupa
                eiw!!.widgetStadium2 = getString(R.string.stadium_2)
                phase123 = false
            }

            5 -> { // Larva
                eiw!!.widgetStadium2 = getString(R.string.stadium_3)
                phase123 = false
            }

            6 -> { // Egg
                eiw!!.widgetStadium2 = getString(R.string.stadium_4)
                phase123 = false
            }
        }
        if (phase123!!) {
            eiw!!.widgetState1(true) // headline status
            eiw!!.setWidgetState1(getString(R.string.status123) + ":")
            eiw!!.widgetState2(true) // state
            eiw!!.setWidgetState2("")
        } else {
            eiw!!.widgetState1(false)
            eiw!!.setWidgetState2("-")
            eiw!!.widgetState2(false)
        }
        eiw!!.setWidgetCount1(getString(R.string.count1) + ":") // icount
        eiw!!.widgetCount2 = 1
        eiw!!.setWidgetIndivNote1(getString(R.string.note) + ":")
        eiw!!.widgetIndivNote2 = ""
        eiw!!.setWidgetXCoord1(getString(R.string.xcoord))
        eiw!!.setWidgetXCoord2(String.format("%.6f", lat))
        eiw!!.setWidgetYCoord1(getString(R.string.ycoord))
        eiw!!.setWidgetYCoord2(String.format("%.6f", lon))
        eiw!!.setWidgetLocality1(getString(R.string.locality) + ":")

        if (metaPref)
            eiw!!.widgetLocality2 = sLocality
        else
            eiw!!.widgetLocality2 = "-"
        indivArea!!.addView(eiw)
    }
    // End of displayData()

    override fun onPause() {
        super.onPause()

        if (IsRunningOnEmulator.DLOG || BuildConfig.DEBUG)
            Log.i(TAG, "259, onPause")

        // Close the data sources
        individualsDataSource!!.close()
        countDataSource!!.close()
    }

    override fun onStop() {
        super.onStop()

        if (IsRunningOnEmulator.DLOG || BuildConfig.DEBUG)
            Log.i(TAG, "270, onStop")

        if (buttonSoundPref)
            soundService!!.releaseSoundP()

        if (awakePref) {
            window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        }
    }

    override fun onDestroy() {
        super.onDestroy()

        if (IsRunningOnEmulator.DLOG || BuildConfig.DEBUG)
            Log.i(TAG, "284, onDestroy")
    }

    private fun saveData(): Boolean {
        // Save individual data
        indivId = individualsDataSource!!.saveIndividual(
            individualsDataSource!!.createIndividuals(
                countId,
                specName,
                lat,
                lon,
                heightNN,
                uncert,
                dateStamp,
                timeStamp,
                eiw!!.widgetLocality2,
                code
            )
        )
        individuals = individualsDataSource!!.getIndividual(indivId)

        // Locality (from reverse geocoding in CountingActivity or manual input)
        individuals!!.locality = eiw!!.widgetLocality2

        // Uncertainty
        if (lat != 0.0) {
            individuals!!.uncert = uncert
        } else {
            individuals!!.uncert = "0"
        }

        // Stadium
        individuals!!.stadium = eiw!!.widgetStadium2

        // State_1-6
        val newState0 = eiw!!.widgetState2
        if (newState0 == "-" || newState0 == "")
            individuals!!.state_1_6 = 0
        else {
            val newState = newState0.toInt()
            if (newState in 0..6) {
                individuals!!.state_1_6 = newState
            } else {
                mesg = getString(R.string.valState)
                Toast.makeText( // orange
                    applicationContext,
                    fromHtml("<font color='#ff6000'><b>$mesg</b></font>"),
                    Toast.LENGTH_LONG
                ).show()
                return false
            }
        }

        // Number of individuals
        val newCount = eiw!!.widgetCount2
        if (newCount > 0) // valid positive newCount
        {
            when (indivAttr) {
                1 -> {
                    // ♂|♀
                    counts!!.count_f1i += newCount
                    individuals!!.icount = newCount
                    individuals!!.sex = "-"
                    individuals!!.icategory = 1
                    countDataSource!!.saveCountf1i(counts!!)
                }

                2 -> {
                    // ♂
                    counts!!.count_f2i += newCount
                    individuals!!.icount = newCount
                    individuals!!.sex = "m"
                    individuals!!.icategory = 2
                    countDataSource!!.saveCountf2i(counts!!)
                }

                3 -> {
                    // ♀
                    counts!!.count_f3i += newCount
                    individuals!!.icount = newCount
                    individuals!!.sex = "f"
                    individuals!!.icategory = 3
                    countDataSource!!.saveCountf3i(counts!!)
                }

                4 -> {
                    // Pupa
                    counts!!.count_pi += newCount
                    individuals!!.icount = newCount
                    individuals!!.sex = "-"
                    individuals!!.icategory = 4
                    countDataSource!!.saveCountpi(counts!!)
                }

                5 -> {
                    // Larva
                    counts!!.count_li += newCount
                    individuals!!.icount = newCount
                    individuals!!.sex = "-"
                    individuals!!.icategory = 5
                    countDataSource!!.saveCountli(counts!!)
                }

                6 -> {
                    // Egg
                    counts!!.count_ei += newCount
                    individuals!!.icount = newCount
                    individuals!!.sex = "-"
                    individuals!!.icategory = 6
                    countDataSource!!.saveCountei(counts!!)
                }
            }

            // Notes
            val newNotes = eiw!!.widgetIndivNote2
            if (newNotes != "") {
                individuals!!.notes = newNotes
            }
            individualsDataSource!!.saveIndividual(individuals)
        } else  // newCount is <= 0
        {
            mesg = getString(R.string.warnCount)
            Toast.makeText( // orange
                this,
                fromHtml("<font color='#ff6000'><b>$mesg</b></font>"),
                Toast.LENGTH_LONG
            ).show()
            return false // forces input newCount > 0
        }
        return true
    }
    // End of saveData()

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.edit_individual, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle action bar item clicks here.
        val id = item.itemId
        if (id == R.id.menuSaveExit) {
            if (saveData()) {
                finish()
            }
            return true
        } else if (id == android.R.id.home) {
            finish()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    private fun buttonVib() {
        if (Build.VERSION.SDK_INT >= 31) {
            vibrator?.vibrate(VibrationEffect.createPredefined(VibrationEffect.EFFECT_CLICK))
        } else {
            if (Build.VERSION.SDK_INT >= 26) {
                vibrator?.vibrate(
                    VibrationEffect.createOneShot(
                        200,
                        VibrationEffect.DEFAULT_AMPLITUDE
                    )
                )
            } else {
                @Suppress("DEPRECATION")
                vibrator?.vibrate(200)
            }
            vibrator?.cancel()
        }
    }

    companion object {
        private const val TAG = "EditIndivAct"
    }

}
